Save draft on chat change

maxphilippov opened the pull request:
- resolves #3733

maxphilippov commented:
This is a thing we discussed a long time ago, we had a 1 second debounce for saving draft, so if you're in any way fast at switching between different chats you just lose what you were typing right before switching. We were talking about testing something around 200-300ms, I went for 100.

While I have a pretty beefy machine (some recent Ryzen 5), I tested this on full power saving both with no attached files and with 200Mb files attached as a draft. I don't see any signs of degradation in how it feels to type. With a 200Mb file attached I have some weird freezes both on `main` and here (if I switch to another contact or sometimes my mouse cursor doesn't move while typing is still perfectly smooth).

So I'm looking for feedback from people who might find a very slow machine and test it on and compare to `main` typing experience. For me 100ms is fast enough no matter how I try I don't lose any draft pieces.

maxphilippov commented:
I did keep the original 1 second delay between autosave

WofWca commented on packages/frontend/src/components/composer/ComposerMessageInput.tsx:
```diff
@@ -29,7 +29,7 @@ export default class ComposerMessageInput extends React.Component<
   setCursorPosition: number | false
   textareaRef: React.RefObject<HTMLTextAreaElement>
-  saveDraft: () => void
+  saveDraft: any
```
```suggestion
  saveDraft: ReturnType<typeof debounce<(text: string, chatId: number) => void>>
```

WofWca commented on CHANGELOG.md:
```diff
@@ -53,6 +53,7 @@
 - cancel old message highlight animations when a new message is highlighted #4203
 - fix: packaging: include architecture in filename for all appimages #4202
 - fix: make open external link scheme case insensive #4201
+- fix: save draft when you switch to another dialog #3733
```
```suggestion
- fix: save draft when you switch to another chat #3733
```

Another chat?

WofWca commented on packages/frontend/src/components/composer/ComposerMessageInput.tsx:
```diff
@@ -124,7 +123,7 @@ export default class ComposerMessageInput extends React.Component<
       if (prevState.text !== this.state.text) {
         if (!this.state.loadingDraft) {
-          this.saveDraft()
+          this.saveDraft.flush(this.state.text, prevState.chatId)
```
`flush()` takes no arguments.

WofWca reviewed (changes requested):
So do you suggest to save the draft on every key press? Because this is what's happening in this MR, and it doesn't seem intentional.

maxphilippov commented:
> So do you suggest to save the draft on every key press? Because this is what's happening in this MR, and it doesn't seem intentional.

Oops, my bad, rushed that a bit. Didn't notice we have a setState above. I'll draft it for now. Reverted back to 200ms, removed saveDraft from onChange then. Gonna look for a way to maybe flush via ref to messageInputComposer. I don't agree with throttle, since I don't see the point in saving first character immediately and thus pushing next save further in time. Next save is what's more likely to happen on chat change.

WofWca commented:
> thus pushing next save further in time

I don't think that's how throttle works. Throttle is never less frequent than debounce.

maxphilippov commented:
As discussed with @nicodh, I've added a simple throttle implementation and made saveDraft use this one instead of debounce.

WofWca commented on packages/shared/util.ts:
```diff
@@ -14,3 +14,30 @@ export function truncateText(text: string, max_len: number) {
+      clearTimeout(timeout)
+      timeout = setTimeout(
+        () => {
+          if (performance.now() - lastTime >= wait) {
```
So if this condition is false, the last call of the throttled function will not invoke `fn`?

Why not use an existing package?

WofWca commented on CHANGELOG.md:
```diff
@@ -13,6 +13,7 @@
 - "Disappearing Messages" dialog not reflecting the actual current value #4327
 - accessibility: make settings keyboard-navigable #4319
 - Fix documentation for --allow-unsafe-core-replacement #4341
+- fix: save draft when you switch to another chat #3733
```
This doesn't sound entirely accurate.

WofWca commented on packages/frontend/src/components/composer/ComposerMessageInput.tsx:
```diff
@@ -124,17 +123,14 @@ export default class ComposerMessageInput extends React.Component<
       if (prevState.text !== this.state.text) {
         if (!this.state.loadingDraft) {
-          this.saveDraft()
+          this.saveDraft(this.state.text, prevState.chatId)
```
Why use `prevState.chatId` here instead of `this.state.chatId`? They're the same inside this `if` block.

WofWca commented on packages/frontend/src/components/composer/ComposerMessageInput.tsx:
```diff
@@ -46,10 +46,9 @@ export default class ComposerMessageInput extends React.Component<
+    this.saveDraft = throttle((text: string, chatId: number) => {
       this.props.updateDraftText(text.trim() === '' ? '' : text, chatId)
-    }, 1000)
+    }, 200)
```
Is the timer supposed to be lowered still?

nicodh replied:
I think this is sufficient at least in my tests. Somehow the saveDraft seems to be blocking anyway so if I click on a chat or account immediately after typing the click has no effect.
I still managed to have a missing character in the draft or one left over after deleting but I think that's ok. I had to type very fast and click mutiple times on another chat to achieve that.

WofWca replied:
> so if I click on a chat or account immediately after typing the click has no effect.

Uhh, is this also on the 1.48 release version?

WofWca replied:
> I think this is sufficient at least in my tests.

What I meant is that is it supposed to be lowered to 200 instead of staying at 1000. Because switching to `throttle` is supposed to be enough.

nicodh replied:
> Uhh, is this also on the 1.48 release version?

Yes with current master.

nicodh replied:
It even blocks my mouse pointer whe I try to move it
On current main
I enter fast 5-6 characters and then try to click somewhere  - it takes until the draft is saved before I can interact again

WofWca commented on packages/frontend/src/components/composer/ComposerMessageInput.tsx:
```diff
@@ -46,10 +46,9 @@ export default class ComposerMessageInput extends React.Component<
 
-    this.saveDraft = debounce(() => {
-      const { text, chatId } = this.state
+    this.saveDraft = throttle((text: string, chatId: number) => {
```
Am I understanding this correctly that the switching from `const { text, chatId } = this.state` to `(text: string, chatId: number) =>` has no functional significance?

WofWca commented on packages/frontend/src/components/composer/ComposerMessageInput.tsx:
```diff
@@ -124,17 +123,14 @@ export default class ComposerMessageInput extends React.Component<
 
   onChange(e: React.ChangeEvent<HTMLTextAreaElement>) {
     this.setState({ text: e.target.value /*error: false*/ })
-    if (!this.state.loadingDraft) {
-      this.saveDraft()
-    }
```
Why is this removed? Is it just a refactor?

nicodh replied:
These lines exist also in onComponentDidUpdate what doesn't make sense to me, since switching to another chat also triggers onComponentDidUpdate so in the end a fresh loaded draft is immediately saved again. So why not remove them there and leave them here?

WofWca replied:
Makes sense.
But the question is if it's directly related to what this commit is doing.
Consider this a nitpick then.

WofWca reviewed (approved):
Tested for a while. Appears to work alright.
I also looked at the `throttle` implementation again, and it appears to be OK for the current purposes.

If the emoji thing is fine, then let's merge.
Hopefully there are no bad side effects in the refactoring part of this MR.
